lib/msprime.c

```diff
@@ -3462,13 +4145,32 @@ msp_insert_root_segments(msp_t *self, const segment_t *head, segment_t **new_hea
             if (ret != 0) {
                 goto out;
             }
-
+            if (self->model.type == MSP_MODEL_SMC_K) {
+                if (self->state == MSP_STATE_NEW) {
```

Comment:
This case isn't tested - what is it we're testing against?

I also wonder about having these type of "model specific" checks deep inside a loop. I think it's better to just do another loop later, just over the individuals (maybe)

Reply:
Good point. I'll restructure the model specific checks.
What I am testing against is the fact that one could instantiate a simulator using the lowlevel or C API with the model directly set to 'smc_k'. However, this is not the logic that is followed when running `sim_ancestry()` for example. Here we always initialise the simulator as if the model was 'hudson' first and only set the model once we start simulating. I have therefore not implemented any smc_k initialisation steps for the non-standard case. The second if-statement catches this cases.

Reply:
The `msp_insert_root_segments` logic only gets called for the SMC_K when we add in historical samples. This approach will only work though if the inserted root segments are sorted based on left coordinate (or if they all have the same left coordinate).